Code/Framework/AzFramework/AzFramework/DocumentPropertyEditor/AdapterBuilder.h

```diff
@@ -0,0 +1,111 @@
+
+namespace AZ::DocumentPropertyEditor
+{
+    //! Helper class that builds a DOM suitable for usage in a DocumentAdapter.
+    //! Uses a visitor pattern to establish node elements.
+    class AdapterBuilder
+    {
+    public:
+        AdapterBuilder() = default;
+
+        //! Begins a new Node, pushing its value to the entry stack.
+        //! Nodes may have child nodes, values, and attributes, as dictated by their node definiton.
+        //! Calls to Attribute and Value will apply to the last BeginNode call.
+        //! Each call to BeginNode must have a corresponding EndNode invocation.
+        void BeginNode(AZ::Name name);
+        //! Ends the current node, popping its value from the entry stack.
+        //! Once the last BeginNode call has had a corresponding EndNode invocation, this builder is in a finished state.
+        //! The result may then be retrieved.
+        void EndNode(AZ::Name expectedName);
+
+        //! Convenience method, calls BeginNode<Nodes::Adapter>()
+        void BeginAdapter();
+        //! Convenience method, calls EndNode<Nodes::Adapter>()
+        void EndAdapter();
+        //! Convenience method, calls BeginNode<Nodes::Row>()
+        void BeginRow();
+        //! Convenience method, calls EndNode<Nodes::Row>()
+        void EndRow();
+        //! Convenience method, calls BeginNode<Nodes::Label>()
+        void BeginLabel();
+        //! Convenience method, calls EndNode<Nodes::Label>()
+        void EndLabel();
+        //! Convenience method, calls BeginNode<Nodes::PropertyEditor>()
+        //! If type and value are specified, the Nodes::PropertyEditor::Type attribute and the
+        //! property editor value shall be set.
+        void BeginPropertyEditor(AZStd::string_view type = "", Dom::Value value = {});
+        //! Convenience method, calls EndNode<Nodes::PropertyEditor>()
+        void EndPropertyEditor();
+
+        //! Inserts a Label node with the specified text.
+        void Label(AZStd::string_view text, bool copy = true);
+
+        //! Sets the value of the last node. Used for setting the current value of a property editor.
+        void Value(Dom::Value value);
+        //! Sets an attribute of the last node. Rows, labels, and property editors all support different attributes.
+        //! \see DocumentPropertyEditor::Nodes
+        void Attribute(Name attribute, Dom::Value value);
+
+        //! Returns true if an error has been encountered during the build process, 
+        bool IsError() const;
+        //! Returns the error information, if any, encountered during the build process.
+        const AZStd::string& GetError() const;
+        //! Ends the build operation and retrieves the builder result.
+        //! Operations are no longer valid on this builder once this is called.
+        Dom::Value&& FinishAndTakeResult();
+
+        template <class NodeDefinition>
+        void BeginNode()
+        {
+            BeginNode(GetNodeName<NodeDefinition>());
+        }
```

Comment:
Nitpick, repeated: A bit of a personal preference, but putting implementations like this in an .inl keep the header more readable.

Reply:
Hmm good callout, I'm going to keep an eye on the size of this - if these get to be larger than one-liners I definitely want to move the implementation out, but right now I think the extra few lines keep the header relatively legible.